### PR TITLE
LockedDB

### PR DESCRIPTION
**Pull request by pinebit**
Closes https://app.shortcut.com/chainlinklabs/story/27740/introduce-locked-db-entity-to-bind-db-conn-and-db-locks

This is the next iteration on "graceful shutdown" effort, here I introduced `LockedDB` entity to bound DB connection and DB locks.

### CODE REVIEW

File: core/cmd/local_client.go [hunk 1]
```diff
@@ -166,14 +109,18 @@ func (cli *Client) runNode(c *clipkg.Context) error {
...
+	// Try opening DB connection and acquiring DB locks at once
+	if err = ldb.Open(rootCtx); err != nil {
+		// If not successful, we know neither locks nor connection remains opened
+		return cli.errorOut(errors.Wrap(err, "opening db"))
 	}
+	defer lggr.ErrorIfClosing(ldb, "db")
 
-	app, err := cli.AppFactory.NewApplication(cli.Config, db)
+	// From now on, DB locks and DB connection will be released on every return.
+	// Keep watching on logger.Fatal* calls and os.Exit(), because defer will not be executed.
```
Comment (samsondav) [inline core/cmd/local_client.go:120]: Lovely comments

File: core/cmd/local_client.go [hunk 2]
```diff
@@ -418,13 +364,13 @@ func (cli *Client) RebroadcastTransactions(c *clipkg.Context) (err error) {
 	}
 
 	lggr := cli.Logger.Named("RebroadcastTransactions")
-	db, err := openDB(cli.Config, lggr)
-	if err != nil {
-		return cli.errorOut(errors.Wrap(err, "opening db"))
+	ldb := pg.NewLockedDB(cli.Config, lggr)
+	if err = ldb.Open(context.Background()); err != nil {
```
Comment (samsondav) [inline core/cmd/local_client.go:368]: Do we want this context to be linked to a sigterm?

File: core/services/pg/locked_db.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Open function connects to DB and acquires DB locks based on configuration.
+// If any of the steps fails or ctx is cancelled, it reverts everything.
+// This is a blocking function and it may execute long due to DB locks acquisition.
+// NOT THREAD SAFE
+func (l *lockedDb) Open(ctx context.Context) (err error) {
+	// If Open succeeded previously, db will not be nil
+	if l.db != nil {
+		return errors.New("calling Open() twice")
```
Comment (samsondav) [inline core/services/pg/locked_db.go:44]: IMO we can panic here, this is a programming error

File: core/services/pg/locked_db.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			return errors.Wrap(err, "failed to take initial lease on database")
+		}
+	}
+
+	// Try to acquire an advisory lock to prevent multiple nodes starting at the same time
+	switch lockingMode {
+	case "advisorylock", "dual":
+		l.advisoryLock = NewAdvisoryLock(l.db, l.cfg.AdvisoryLockID(), l.lggr, l.cfg.AdvisoryLockCheckInterval())
+		if err = l.advisoryLock.TakeAndHold(ctx); err != nil {
+			defer revert()
```
Comment (samsondav) [inline core/services/pg/locked_db.go:79]: revert should include undoing the lease lock

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/pg/locked_db.go (hunk 4 above)
Comment (pinebit) [reply core/services/pg/locked_db.go:79]: `revert()` calls `LockDB.Close()` which in turn releases all locks and closes DB.

File: core/cmd/local_client.go (hunk 2 above)
Comment (pinebit) [reply core/cmd/local_client.go]: Not relevant anymore, given the new API.

**Review by samsondav — APPROVED** [review]

**Pull request merged**